homeassistant/components/sensor/aemet.py

```diff
@@ -0,0 +1,191 @@
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+    name = config.get(CONF_NAME)
+    api_key = config.get(CONF_API_KEY)
+    station_id = config.get(CONF_STATION_ID)
+
+    aemet_data = AemetData(api_key=api_key, station_id=station_id)
+    try:
+        aemet_data.update()
+    except (ValueError, TypeError) as err:
+        _LOGGER.error("Received error from AEMET: %s", err)
+        return False
+
+    add_entities([AemetSensor(aemet_data, variable, name)
+                  for variable in config[CONF_MONITORED_CONDITIONS]], True)
+
+
+class AemetSensor(Entity):
+    """Representation of a sensor in the AEMET service."""
+
+    def __init__(self, aemet_data, variable, name):
+        """Initialize the sensor."""
+        self.aemet_data = aemet_data
+        self.variable = variable
+        self.client_name = name
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return '{} {}'.format(self.client_name, self.variable)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self.aemet_data.get_data(self.variable)
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return SENSOR_TYPES[self.variable][1]
+
+    @property
+    def icon(self):
+        """Return sensor specific icon."""
+        return SENSOR_TYPES[self.variable][2]
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            ATTR_STATION_NAME: self.aemet_data.get_data(ATTR_STATION_NAME),
+            ATTR_LATITUDE: self.aemet_data.get_data(ATTR_LATITUDE),
+            ATTR_LONGITUDE: self.aemet_data.get_data(ATTR_LONGITUDE),
+            ATTR_ELEVATION: self.aemet_data.get_data(ATTR_ELEVATION),
+            ATTR_LAST_UPDATE: self.aemet_data.get_data(ATTR_LAST_UPDATE),
+        }
+
+    def update(self):
+        """Delegate update to data class."""
+        self.aemet_data.update()
+
+
+class AemetData:
+    """Get the lastest data and updates the states."""
+
+    API_URL_BASE = 'https://opendata.aemet.es/opendata/api'
+    API_STATION_ENDPOINT = '/observacion/convencional/datos/estacion/{}'
+
+    def __init__(self, api_key, station_id):
+        """Initialize the data object."""
+        self._station_id = station_id
+        self._api_key = api_key
+        self.data = {}
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Fetch new state data for the sensor."""
+        _LOGGER.debug("------- Updating AEMET sensor")
+
+        endpoint_url = "{}{}".format(
+            self.API_URL_BASE,
+            self.API_STATION_ENDPOINT.format(self._station_id)
+        )
+        params = {'api_key': self._api_key}
+        main_rsp = requests.get(endpoint_url, params=params)
```

Comment:
you should not reach out to an API from HA. Data should come from an external pypi package. See https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
Added changes to avoid requests from HA by using the pypi library `python-aemet` but I found a bug in that library, so my code is pending of this PR https://github.com/pablo-moreno/python-aemet/pull/5 in order to work fine.

Reply:
This component now reach the aemet API using the pypi library `python-aemet==0.1.4`

Reply:
@uchagani please take another look